File: app/frontend/modules/survey/models/Validations/Exact.js
import extend from 'lodash/extend'
import { I18n } from 'modules/survey/store/StoreWatchman'

const Exact = function ({ exactValue }) {
  this.exactValue = +exactValue
}

extend(Exact.prototype, {
  validate (answers) {
    if (answers.length !== this.exactValue) {
      return {
        type: 'Exact',
        message: I18n().t('validations.exact', { count: this.exactValue }),
      }
    }
  },
})

export default Exact

File: app/core/questions/validations/exact.rb
# frozen_string_literal: true

module Questions::Validations
  class Exact < BaseCommand
    attr_accessor :args, :message, :value

    def initialize(validation, value, locale)
      @args = validation['args']
      @value = value || []
      @locale = locale
    end

    def call
      return broadcast :ok, nil unless args['exactValue']

      invalid = value.length > args['exactValue']

      broadcast :ok, invalid ? [I18n.t('validations.exact', count: args['exactValue'])] : nil
    end
  end
end
